File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq


def solution(operations):
    max_hq, min_hq, removed = [], [], set()
    for operation in operations:
        inst, num = operation.split()
        num = int(num)
        if inst == 'D':
            if num == 1:
                while max_hq and -max_hq[0] in removed:
                    heapq.heappop(max_hq)
                if max_hq:
                    removed.add(-heapq.heappop(max_hq))
            else:
                while min_hq and min_hq[0] in removed:
                    heapq.heappop(min_hq)
                if min_hq:
                    removed.add(heapq.heappop(min_hq))
        else:
            heapq.heappush(min_hq, num)
            heapq.heappush(max_hq, -num)
    while min_hq and min_hq[0] in removed:
        heapq.heappop(min_hq)
    while max_hq and -max_hq[0] in removed:
        heapq.heappop(max_hq)
    return [-heapq.heappop(max_hq) if max_hq else 0, heapq.heappop(min_hq) if min_hq else 0]

File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minPq = new PriorityQueue<>();
        PriorityQueue<Integer> maxPq = new PriorityQueue<>(Collections.reverseOrder());
        Set<Integer> deleted = new HashSet<>();
        int[] answer = {0, 0};

        for (String operation: operations){
            String[] inst = operation.split(" ");
            int num = Integer.parseInt(inst[1]);

            if (inst[0].equals("D")){
                if (num == 1){
                    while (!maxPq.isEmpty() && deleted.contains(maxPq.peek()))
                        maxPq.poll();
                    if (!maxPq.isEmpty()) {
                        deleted.add(maxPq.poll());
                    }
                } else {
                    while (!minPq.isEmpty() && deleted.contains(minPq.peek()))
                        minPq.poll();
                    if (!maxPq.isEmpty()) {
                        deleted.add(minPq.poll());
                    }
                }
            } else {
                maxPq.add(num);
                minPq.add(num);
            }
        }

        while (!maxPq.isEmpty() && deleted.contains(maxPq.peek()))
            maxPq.poll();
        while (!minPq.isEmpty() && deleted.contains(minPq.peek()))
            minPq.poll();

        answer[0] = maxPq.isEmpty() ? 0 : maxPq.peek();
        answer[1] = minPq.isEmpty() ? 0 : minPq.peek();
        return answer;
    }
}
